### ISSUE TITLE
How to get Sentence embedding using pre-trained SciBERT weights? 

### ISSUE THREAD

**Issue by Santosh-Gupta** [allenai/scibert#34]
The instructions to use SciBERT say this

> SciBERT models include all necessary files to be plugged in your own model and are in same format as BERT. If you are using Tensorflow, refer to Google's BERT repo and if you use PyTorch, refer to Hugging Face's repo where detailed instructions on using BERT models are provided.

However, to use BERT, the instructions include loading BERT from tf.hub. But I don't see SciBert on tfhub, so I am unable to figure out how to get a sentence embedding for SciBert. 

This is my attempt in Python. So far I cloned the repository and loaded the weights, but I don't know how to get the sentence/paragraph vector. 

This is my code so far

```
import sys
import tensorflow as tf

!test -d SciBert_repo || git clone https://github.com/allenai/scibert SciBert_repo 
if not 'SciBert_repo ' in sys.path:
  sys.path += ['SciBert_repo ']

import extract_features

with tf.Session(graph=graph) as session:
 
   saver.restore(session, 'SciBert_repo.ckpt' )
```

I know this is based on the original BERT code. For regular BERT they have you use tf.hub, but I'm guessing the setup is pretty similar. This is my code for regular BERT

```
pip install bert-tensorflow

import tensorflow as tf
import tensorflow_hub as hub

import bert
from bert import run_classifier
from bert import optimization
from bert import tokenization

import pandas as pd

from tensorflow import keras
import os
import re

from tensorflow.keras import backend as K

from bert.tokenization import FullTokenizer

bert_path = "https://tfhub.dev/google/bert_uncased_L-12_H-768_A-12/1"

sess = tf.Session()

bert_module = hub.Module(
  bert_path,
  trainable=True)

#Basically this is a function to convert text into a format BERT understands
def bertInputsFromText(text):
.
.
.

bert_inputs = bertInputsFromText("This is a test sentence")

sentence_embedding= bert_module(inputs=bert_inputs, signature="tokens", as_dict=True)[
        "pooled_output"
```

So I'm guessing my question boils down to, what to use as an equivalent to `bert_module `

**Comment by Santosh-Gupta** [comment]
I was able to do this by using Bert as a Service

https://github.com/hanxiao/bert-as-service

And loading the pre-trained weights there.

**Comment by Saichethan** [comment]
Hello @Santosh-Gupta , how to get sentence embedding using pre trained SciBert weights?

**Comment by ibeltagy** [comment]
Depending on the application, you can use the embedding corresponding to the CLS token (which I would do) or pooling embedding of all the words. 
If you want to embed a paper specifically, you might want to consider [SPECTER](https://github.com/allenai/specter) as well.

**Comment by FL33TW00D** [comment]
> 
> 
> Depending on the application, you can use the embedding corresponding to the CLS token (which I would do) or pooling embedding of all the words.
> If you want to embed a paper specifically, you might want to consider [SPECTER](https://github.com/allenai/specter) as well.

Hi @ibeltagy,
I am digging into using SciBert for sentence embeddings and stumbled upon your comment, I am curious as to why you suggested using the CLS token as the sentence representation as opposed to pooling the embedding of the word vectors.

![image](https://user-images.githubusercontent.com/45471420/108044729-8a8f6700-703a-11eb-9723-0d4c22f751f4.png)

From the S-BERT paper it seems there is a big performance difference between CLS and pooling.

Many Thanks,
Chris
